### PR TITLE
add option to force re-authentication for OIDC connectors

### PR DESCRIPTION
**Pull request by capnspacehook**
Protobuf changes mostly, most of the implementation is here: https://github.com/gravitational/teleport.e/pull/1915

### CODE REVIEW

**Review by EdwardDowling — APPROVED** [review]

File: api/proto/teleport/legacy/types/types.proto
Comment (zmb3) [inline api/proto/teleport/legacy/types/types.proto:3930]: Should we just expose this as `MaxAge` instead?

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto:3930]: I wasn't sure, we certainly can but I was just erring on the side of simplicity for the user as no one's asked for this before

Comment (zmb3) [reply api/proto/teleport/legacy/types/types.proto:3930]: Yeah, I can see that argument. But sometimes for things as complicated as SSO it makes more sense to expose the field directly rather than creating some abstraction over it.

People familiar with the OIDC flow might know to search for `max_age` but they probably won't know that we hid that option behind something called `force_reauthentication`.

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto]: I've been trying to think of a simple way to make `max_age` nullable, but I'm not sure how to do that with protobufs. Apparently `optional` is off the table, and `gogoproto.nullable` doesn't seem to change the representation of `max_age`. This matters because the zero-value for `max_age` is a valid value, but a user may not want `max_age` set at all to keep existing behavior.

I toyed with the idea of setting a default value for `max_age` that was very large, but that would technically change behavior silently and be backwards incompatible. Setting `max_age` to even a large value would be more secure than not setting it at all, but I'm not sure how important it is to keep behavior *exactly* the same if a user upgrades Teleport in this case. WDYT @zmb3?

File: api/proto/teleport/legacy/types/types.proto
Comment (r0mant) [inline api/proto/teleport/legacy/types/types.proto:3929]: Update the godoc.

**Review by r0mant — APPROVED** [review]

File: api/proto/teleport/legacy/types/types.proto
Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto]: Another (ugly) option is to introduce a second field, ie `max_age_enabled` but that seems very unintuitive

Comment (r0mant) [reply api/proto/teleport/legacy/types/types.proto]: I think we need to bump the resource version to keep this backwards compatible. So the current version of the connector will not support max_age parameter and v4 will.

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto]: Fixed with help from @codingllama

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto]: I figured out a way to make `max_age` nullable, so existing connectors will default to have it set to `nil` and after updating users can set `max_age` to a concrete value if they wish. This new reauth behavior will only take effect when `max_age` is set. Not sure how that's backwards incompatible?

Comment (r0mant) [reply api/proto/teleport/legacy/types/types.proto]: Yeah that should be fine then. Just make sure to test it please.

File: api/proto/teleport/legacy/types/types.proto
Comment (tigrato) [inline api/proto/teleport/legacy/types/types.proto:3943]: Given that the field is a duration why don't you use the well known types:

```suggestion
  google.protobuf.Duration MaxAge = 17 [
    (gogoproto.jsontag) = "max_age",
    (gogoproto.stdduration) = true
  ];
}
```

This is compiled to:

```go
MaxAge *time.Duration
```

This also fixes your operator issues

Comment (tigrato) [reply api/proto/teleport/legacy/types/types.proto:3943]: and the operator transforms it to `string` instead of `int32`

```
              max_age:
                description: MaxAge is the amount of time that user logins are valid
                  for. If a user logs in, but then does not login again within this
                  time period, they will be forced to re-authenticate.
                format: duration
                type: string
```
---

**Having a duration with `int32` is a HUGE problem because int32 can only encode durations up to `2.14s`**

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto:3943]: I'm weary of using `(gogoproto.stdduration) = true` instead of `(gogoproto.casttype) = "Duration"`, as all existing code uses `Duration`. I'm not 100% sure why, but I'm sure there's a good reason for it. Maybe the CRD just needs to be updated to handle this case?

File: api/types/oidc.go
Comment (codingllama) [inline api/types/oidc.go:438]: This implies MaxAge is in nanos. Is that what you wanted?

File: api/proto/teleport/legacy/types/types.proto
Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto:3966]: A `api/types.Duration` is the underlying type, see https://github.com/gravitational/teleport/pull/29815/files#diff-baa2be43d621c39aa299dbc24f5f09e6ec5e75a4c8a12202112856abf3d10cc3R3976.

I tried using `google.protobuf.Duration`, but it broke marshalling resource YAML data in the web UI.

Comment (codingllama) [reply api/proto/teleport/legacy/types/types.proto:3966]: So the Duration is in nanos, is that correct? Please document it.

Comment (capnspacehook) [reply api/proto/teleport/legacy/types/types.proto:3966]: Done

**Review by codingllama — COMMENTED** [review]
LGTM, minus CI failures.

**Comment by capnspacehook** [comment]
Those will get resolved once I get a G1 approval for https://github.com/gravitational/teleport/pull/29916 and merge the changes into here

**Pull request merged**
